src/ddmd/root/file.d

```diff
@@ -77,12 +77,69 @@ nothrow:
     {
         if (len)
             return false; // already read the file
+
+        import core.stdc.string : strcmp;
+        const(char)* name = this.name.toChars();
+        if (strcmp(name, "-.d") == 0)
+        {
+            /* Read from stdin */
+            import ddmd.root.rmem : mem;
+
+            enum bufIncrement = 128 * 1024;
+            size_t pos = 0;
+            size_t sz = bufIncrement;
+
+            if (!_ref)
+                .free(buffer);
+
+            buffer = null;
+            L1: for (;;)
+            {
+                buffer = cast(ubyte*).realloc(buffer, sz + 2); // +2 for sentinel
+                if (!buffer)
+                {
+                    printf("\tmalloc error, errno = %d\n", errno);
+                    goto err3;
+                }
+
+                // Fill up buffer
+                do
+                {
+                    size_t rlen;
+
+                    assert(sz > pos);
+                    pos += rlen = fread(buffer + pos, 1, sz - pos, stdin);
+                    if (ferror(stdin))
+                    {
+                        printf("\tread error, errno = %d\n", errno);
+                        goto err3;
+                    }
+                    if (feof(stdin))
+                    {
+                        // We're done
+                        assert(pos < sz + 2);
+                        len = pos;
+                        buffer[pos] = '\0';
+                        buffer[pos+1] = '\0';
+                        return false;
+                    }
+                } while (pos < sz);
+
+                // Buffer full, expand
+                sz += bufIncrement;
+            }
+        err3:
+            .free(buffer);
+            buffer = null;
+            len = 0;
+            return true;
+        }
```

Comment:
(Not familiar with the codebase.) Why couldn't the existing file reading code be used?